packages/samples/mvp-dapp/src/exception.ts

```diff
@@ -0,0 +1,21 @@
+import { Context, HttpError, Next } from 'koa'
+
+export function exceptionHandle() {
+  return async (ctx: Context, next: Next) => {
+    try {
+      await next()
+    } catch (err) {
+      if (err instanceof HttpError) {
+        ctx.body = { message: err.message }
+      } else if (err instanceof BizError) {
+        ctx.status = 400
+        ctx.body = { message: err.message }
+      } else {
+        ctx.status = 500
+        ctx.body = { message: 'Internal server error' }
+      }
+    }
+  }
+}
+
+export class BizError extends Error {}
```

Comment:
What if adding a biz error code in the biz error so we can make the response clear, as
HTTP code 404 => endpoint is not found
Biz code 404 => store/field is not found

Reply:
> What if adding a biz error code in the biz error so we can make the response clear, as
> HTTP code 404 => endpoint is not found
> Biz code 404 => store/field is not found

That's what I was thinking, using `MvpResponse` to wrap the return value and business error, it would look something like this

```js
// definition:

export class BizError extends Error {
  errCode: string;

  constructor(message: string, errCode: string) {
    super(message);
    this.errCode = errCode;
  }
}

export class MvpResponse {
  code: string
  data: any;

  constructor(code: string, data: any) {
    this.code = code;
    this.data = data;
  }

  static ok(data: any) {
    return new MvpResponse(data, "200")
  }

  static err(data: any, code: string) {
    return new MvpResponse(data, code)
  }
}

// controller:

router.get<never, { path: string; address: string }>('/load/:address/:path', async (ctx) => {
  const { params } = ctx.payload

  if (!params?.path) {
    throw new BadRequest('invalid path')
  }

  const recordModel = await getRecordModel(params?.address)
  const value = recordModel.load(`data.${params.path}`)
  if (value) {
    ctx.ok(MvpResponse.ok(value))
  } else {
    throw new BizError('field is not found', '404')
    // or
    // ctx.ok(MvpResponse.err("field is not found", '404'))
  }
})

// err handle
export function exceptionHandle() {
  return async (ctx: Context, next: Next) => {
    try {
      await next()
    } catch (err) {
      if (err instanceof HttpError) {
        ctx.body = { message: err.message }
      } else if (err instanceof BizError) {
        ctx.status = 200
        ctx.body = MvpResponse.err(err.message, err.errCode);
      } else {
        ctx.status = 500
        ctx.body = { message: 'Internal server error' }
      }
    }
  }
}
```
Here httpCode is used for HTTP errors, 500 for the rest of the errors, and a code is used in the return value to explain the business situation

Reply:
> > What if adding a biz error code in the biz error so we can make the response clear, as
> > HTTP code 404 => endpoint is not found
> > Biz code 404 => store/field is not found
> 
> That's what I was thinking, using `MvpResponse` to wrap the return value and business error, it would look something like this
> 
> ```js
> // definition:
> 
> export class BizError extends Error {
>   errCode: string;
> 
>   constructor(message: string, errCode: string) {
>     super(message);
>     this.errCode = errCode;
>   }
> }
> 
> export class MvpResponse {
>   code: string
>   data: any;
> 
>   constructor(code: string, data: any) {
>     this.code = code;
>     this.data = data;
>   }
> 
>   static ok(data: any) {
>     return new MvpResponse(data, "200")
>   }
> 
>   static err(data: any, code: string) {
>     return new MvpResponse(data, code)
>   }
> }
> 
> // controller:
> 
> router.get<never, { path: string; address: string }>('/load/:address/:path', async (ctx) => {
>   const { params } = ctx.payload
> 
>   if (!params?.path) {
>     throw new BadRequest('invalid path')
>   }
> 
>   const recordModel = await getRecordModel(params?.address)
>   const value = recordModel.load(`data.${params.path}`)
>   if (value) {
>     ctx.ok(MvpResponse.ok(value))
>   } else {
>     throw new BizError('field is not found', '404')
>     // or
>     // ctx.ok(MvpResponse.err("field is not found", '404'))
>   }
> })
> 
> // err handle
> export function exceptionHandle() {
>   return async (ctx: Context, next: Next) => {
>     try {
>       await next()
>     } catch (err) {
>       if (err instanceof HttpError) {
>         ctx.body = { message: err.message }
>       } else if (err instanceof BizError) {
>         ctx.status = 200
>         ctx.body = MvpResponse.err(err.message, err.errCode);
>       } else {
>         ctx.status = 500
>         ctx.body = { message: 'Internal server error' }
>       }
>     }
>   }
> }
> ```
> 
> Here httpCode is used for HTTP errors, 500 for the rest of the errors, and a code is used in the return value to explain the business situation

Upvote